5/assets/5.1-finished-code.rs

```diff
@@ -0,0 +1,307 @@
+
+        let new_owned_kitty_count_from = owned_kitty_count_from.checked_sub(1)
+            .ok_or("Transfer causes underflow of 'from' kitty balance")?;
+
+        let kitty_index = <OwnedKittiesIndex<T>>::get(kitty_id);
+        if kitty_index != new_owned_kitty_count_from {
+            let last_kitty_id = <OwnedKittiesArray<T>>::get((from.clone(), new_owned_kitty_count_from));
+            <OwnedKittiesArray<T>>::insert((from.clone(), kitty_index), last_kitty_id);
+            <OwnedKittiesIndex<T>>::insert(last_kitty_id, kitty_index);
+        }
+
+        <KittyOwner<T>>::insert(&kitty_id, &to);
+        <OwnedKittiesIndex<T>>::insert(kitty_id, owned_kitty_count_to);
+
+        <OwnedKittiesArray<T>>::remove((from.clone(), new_owned_kitty_count_from));
+        <OwnedKittiesArray<T>>::insert((to.clone(), owned_kitty_count_to), kitty_id);
+
+        <OwnedKittiesCount<T>>::insert(&from, new_owned_kitty_count_from);
+        <OwnedKittiesCount<T>>::insert(&to, new_owned_kitty_count_to);
+        
+        Self::deposit_event(RawEvent::Transferred(from, to, kitty_id));
+        
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+	// Import a bunch of dependencies from substrate core. All needed for some parts of the code.
+	use support::{impl_outer_origin, assert_ok, assert_noop};
+	use runtime_io::{with_externalities, TestExternalities};
+	use primitives::{H256, Blake2Hasher};
+	use runtime_primitives::{
+		BuildStorage,
+        traits::{BlakeTwo256, IdentityLookup},
+		testing::{Digest, DigestItem, Header}
+	};
+
+	impl_outer_origin! {
+		pub enum Origin for KittiesTest {}
+	}
+
+	// Create one configuration type (`Test`) which `impl`s each of the
+	// configuration traits of modules we want to use.
+	#[derive(Clone, Eq, PartialEq)]
+	pub struct KittiesTest;
+
+	// Start implementing the Trait's of all the other modules that you need.
+	// If you want anything that is reasonably functional you also need to implement the System trait.
+	impl system::Trait for KittiesTest {
+		type Origin = Origin;
+		type Index = u64;
+		type BlockNumber = u64;
+		type Hash = H256;
+		type Hashing = BlakeTwo256;
+		type Digest = Digest;
+		type AccountId = u64;
+		type Lookup = IdentityLookup<Self::AccountId>;
+		type Header = Header;
+		type Event = ();
+		type Log = DigestItem;
+	}
+	
+    // And any other trait that your Trait is explicitly bounded by.
+	// Remember you had: `pub trait Trait: balances::Trait`
+	impl balances::Trait for KittiesTest {
+		type Balance = u64;
+		type OnFreeBalanceZero = ();
+		type OnNewAccount = ();
+		type Event = ();
+		type TransactionPayment = ();
+		type TransferPayment = ();
+		type DustRemoval = ();
+	}
+
+	// And finally, your own trait.
+	impl super::Trait for KittiesTest {
+		type Event = ();
+	}
+
+	// Creates aliases for modules for easier access
+	type Kitties = super::Module<KittiesTest>;
+	type Balances = balances::Module<KittiesTest>;
+	type System = system::Module<KittiesTest>;
```

Comment:
So as expected this is not really used :D neither is the balances Trait. We can either:
- Let them be as a reference for how to do it. I don't like the compiler warnings though.
- Even better, on the first `ACTION: write some kind of an assertion`, we can also ask them to do one assertion from balances and system as well, or just play around with them.